File: original/modules/music_player.py
# 音乐播放模块
import asyncio
import random
from typing import List, Dict, Any

class MusicPlayer:
    """音乐播放器类"""
    
    def __init__(self):
        self.playlist: List[Dict[str, str]] = []  # 播放列表
        self.current_song = None
        self.play_mode = "album"  # album, single, repeat, loop
        self.is_playing = False
        
    def add_to_playlist(self, title: str, url: str, singer: str = ""):
        """添加到播放列表"""
        song = {
            "title": title,
            "url": url,
            "singer": singer
        }
        self.playlist.append(song)
        return f"已添加到播放列表: {self.format_song_title(song)}"
        
    def remove_from_playlist(self, index: int):
        """从播放列表移除"""
        if 0 <= index < len(self.playlist):
            song = self.playlist.pop(index)
            return f"已从播放列表移除: {self.format_song_title(song)}"
        else:
            return "索引超出范围"
            
    def list_playlist(self):
        """列出播放列表"""
        if not self.playlist:
            return "播放列表为空"
            
        result = "播放列表:\n"
        for i, song in enumerate(self.playlist):
            result += f"{i+1}. {self.format_song_title(song)}\n"
        return result.strip()
        
    def format_song_title(self, song: Dict[str, str]):
        """格式化歌曲标题"""
        title = song["title"]
        if song["singer"]:
            title = f"{title} - {song['singer']}"
        return title
        
    def set_play_mode(self, mode: str):
        """设置播放模式"""
        valid_modes = ["album", "single", "repeat", "loop"]
        if mode in valid_modes:
            self.play_mode = mode
            return f"播放模式已设置为: {mode}"
        else:
            return f"无效的播放模式。可用模式: {', '.join(valid_modes)}"
            
    def get_next_song(self):
        """获取下一首歌曲"""
        if not self.playlist:
            return None
            
        if self.play_mode == "single":
            # 单曲播放，不自动播放下一首
            return None
        elif self.play_mode == "repeat":
            # 单曲循环，返回当前歌曲
            return self.current_song
        elif self.play_mode == "loop":
            # 列表循环，返回第一首
            return self.playlist[0]
        else:
            # 专辑模式，按顺序播放
            if self.current_song and self.current_song in self.playlist:
                current_index = self.playlist.index(self.current_song)
                if current_index + 1 < len(self.playlist):
                    return self.playlist[current_index + 1]
                else:
                    return None  # 播放完毕
            else:
                return self.playlist[0]  # 从第一首开始
                
    def play_next(self):
        """播放下一首"""
        next_song = self.get_next_song()
        if next_song:
            self.current_song = next_song
            self.is_playing = True
            # 从播放列表移除（如果不是循环模式）
            if self.play_mode != "repeat" and self.play_mode != "loop":
                if next_song in self.playlist:
                    self.playlist.remove(next_song)
            return next_song
        else:
            self.is_playing = False
            return None
            
    def shuffle_playlist(self):
        """随机播放列表"""
        if self.playlist:
            random.shuffle(self.playlist)
            return "播放列表已随机排序"
        else:
            return "播放列表为空"
            
    def clear_playlist(self):
        """清空播放列表"""
        self.playlist.clear()
        self.current_song = None
        self.is_playing = False
        return "播放列表已清空"

File: android/app/src/main/java/com/drrr/bot/MusicPlayer.java
package com.drrr.bot;

import android.util.Log;
import java.util.List;
import java.util.ArrayList;
import java.util.Map;
import java.util.HashMap;

public class MusicPlayer {
    private static final String TAG = "MusicPlayer";
    
    // 播放列表
    private List<Map<String, String>> playlist;
    
    // 当前播放的歌曲
    private Map<String, String> currentSong;
    
    // 播放模式 (album, single, repeat, loop)
    private String playMode = "album";
    
    // 播放状态
    private boolean isPlaying = false;
    
    public MusicPlayer() {
        playlist = new ArrayList<>();
        Log.d(TAG, "MusicPlayer initialized");
    }
    
    /**
     * 添加歌曲到播放列表
     * @param title 歌曲标题
     * @param url 歌曲链接
     * @param singer 歌手（可选）
     * @return 结果消息
     */
    public String addToPlaylist(String title, String url, String singer) {
        Map<String, String> song = new HashMap<>();
        song.put("title", title);
        song.put("url", url);
        song.put("singer", singer != null ? singer : "");
        
        playlist.add(song);
        return "已添加到播放列表: " + formatSongTitle(song);
    }
    
    /**
     * 从播放列表移除歌曲
     * @param index 歌曲索引
     * @return 结果消息
     */
    public String removeFromPlaylist(int index) {
        if (index >= 0 && index < playlist.size()) {
            Map<String, String> song = playlist.remove(index);
            return "已从播放列表移除: " + formatSongTitle(song);
        } else {
            return "索引超出范围";
        }
    }
    
    /**
     * 列出播放列表
     * @return 播放列表内容
     */
    public String listPlaylist() {
        if (playlist.isEmpty()) {
            return "播放列表为空";
        }
        
        StringBuilder result = new StringBuilder("播放列表:\n");
        for (int i = 0; i < playlist.size(); i++) {
            result.append(i + 1).append(". ").append(formatSongTitle(playlist.get(i))).append("\n");
        }
        return result.toString().trim();
    }
    
    /**
     * 格式化歌曲标题
     * @param song 歌曲信息
     * @return 格式化后的标题
     */
    private String formatSongTitle(Map<String, String> song) {
        String title = song.get("title");
        String singer = song.get("singer");
        
        if (singer != null && !singer.isEmpty()) {
            title = title + " - " + singer;
        }
        return title;
    }
    
    /**
     * 设置播放模式
     * @param mode 播放模式
     * @return 结果消息
     */
    public String setPlayMode(String mode) {
        if (isValidPlayMode(mode)) {
            this.playMode = mode;
            return "播放模式已设置为: " + mode;
        } else {
            return "无效的播放模式: " + mode;
        }
    }
    
    /**
     * 检查播放模式是否有效
     * @param mode 播放模式
     * @return true if valid, false otherwise
     */
    private boolean isValidPlayMode(String mode) {
        return "album".equals(mode) || "single".equals(mode) || 
               "repeat".equals(mode) || "loop".equals(mode);
    }
    
    /**
     * 播放下一首歌曲
     * @return 结果消息
     */
    public String playNext() {
        if (playlist.isEmpty()) {
            return "播放列表为空";
        }
        
        // 简单实现：播放第一首歌曲
        currentSong = playlist.get(0);
        isPlaying = true;
        return "正在播放: " + formatSongTitle(currentSong);
    }
    
    /**
     * 停止播放
     * @return 结果消息
     */
    public String stop() {
        currentSong = null;
        isPlaying = false;
        return "已停止播放";
    }
    
    /**
     * 清空播放列表
     * @return 结果消息
     */
    public String clearPlaylist() {
        playlist.clear();
        currentSong = null;
        isPlaying = false;
        return "播放列表已清空";
    }
    
    // Getter方法
    public boolean isPlaying() {
        return isPlaying;
    }
    
    public Map<String, String> getCurrentSong() {
        return currentSong;
    }
    
    public String getPlayMode() {
        return playMode;
    }
    
    public int getPlaylistSize() {
        return playlist.size();
    }
}